### PR TITLE
[PodSecurity] Add privileged containers baseline check

### LINKED ISSUE #103196: [PodSecurity] baseline - privileged containers
Implement the "Privileged" check from https://kubernetes.io/docs/concepts/security/pod-security-standards/#baseline

- [ ] implement check_privileged.go for the baseline policy
  - fail if any container sets `privileged` to true, and include the names of the invalid containers in the forbidden details
  - see staging/src/k8s.io/pod-security-admission/policy/check_runAsNonRoot.go as an example
- [ ] implement fixtures_privileged.go
  - see staging/src/k8s.io/pod-security-admission/test/fixtures_runAsNonRoot.go as an example
  - generatePass can just return the "minimal valid pod" input
  - generateFail should generate two test cases with privileged in a container and in an initContainer
  - run `UPDATE_POD_SECURITY_FIXTURE_DATA=true go test k8s.io/pod-security-admission/test` to update test fixtures
  - run `go test ./test/integration/auth -v -run PodSecurity` to ensure integration tests pass

### PR DESCRIPTION
**Pull request by aramase**
Signed-off-by: Anish Ramasekar <anish.ramasekar@gmail.com>

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
- Adds baseline check for privileged containers

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/103196

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
[KEP] https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/2579-psp-replacement
```

### CODE REVIEW

File: staging/src/k8s.io/pod-security-admission/policy/check_privileged.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	visitContainersWithPath(podSpec, field.NewPath("spec"), func(container *corev1.Container, path *field.Path) {
+		if container.SecurityContext != nil && container.SecurityContext.Privileged != nil && *container.SecurityContext.Privileged {
+			privilegedContainers = append(privilegedContainers, container.Name)
+		}
+	})
+	if len(privilegedContainers) > 0 {
+		return CheckResult{
+			Allowed:         false,
+			ForbiddenReason: "privileged == true",
+			ForbiddenDetail: strings.Join(privilegedContainers, ", "),
```
Comment (tallclair) [inline staging/src/k8s.io/pod-security-admission/policy/check_privileged.go:68]: We need some better guidelines on what the reason & details should look like. For now, let's use the field paths for the details to be consistent with the runAsNonRoot check: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/pod-security-admission/policy/check_runAsNonRoot.go#L81

File: staging/src/k8s.io/pod-security-admission/test/fixtures_privileged.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+containerFields: []string{
+	`securityContext.privileged`,
+},
+
+*/
+
+func init(){
+	fixtureData_1_0 := fixtureGenerator{
+		generatePass: func(p *corev1.Pod) []*corev1.Pod {
+			return []*corev1.Pod{p}
```
Comment (tallclair) [inline staging/src/k8s.io/pod-security-admission/test/fixtures_privileged.go:37]: The minimal pod is always included, no need to add it explicitly.

Comment (tallclair) [inline staging/src/k8s.io/pod-security-admission/test/fixtures_privileged.go:37]: Add a test case for explicitly setting privileged to false.

**Review by tallclair — COMMENTED** [review]
looks good, just a couple minor comments.

**Comment by aramase** [comment]
@tallclair Thank you for the review. Updated the forbidden reason and updated fixtures based on the comments.

File: staging/src/k8s.io/pod-security-admission/test/fixtures_privileged.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+			p = ensureSecurityContext(p)
+			return []*corev1.Pod{
+				// privileged set to false in container
+				tweak(p, func(p *corev1.Pod) {
+					p.Spec.Containers[0].SecurityContext.Privileged = pointer.BoolPtr(false)
+				}),
+				// privileged set to false in init container
+				tweak(p, func(p *corev1.Pod) {
+					p.Spec.InitContainers[0].SecurityContext.Privileged = pointer.BoolPtr(false)
+				}),
```
Comment (liggitt) [inline staging/src/k8s.io/pod-security-admission/test/fixtures_privileged.go:46]: combine these to a single test case (that minimizes number of fixtures without losing coverage) and regenerate testdata

**Review by liggitt — COMMENTED** [review]
one comment on the testcases, then lgtm

**Comment by liggitt** [comment]
oh, looks like you also need to update the integration test in test/integration/auth/podsecurity_test.go to pass `--allow-privileged=true` to even get past validation to exercise this admission check for privileged pods

**Comment by liggitt** [comment]
aaaand `hack/update-gofmt.sh` needs to be run :)

we serve the CI robots, clearly :)

**Comment by aramase** [comment]
> oh, looks like you also need to update the integration test in test/integration/auth/podsecurity_test.go to pass --allow-privileged=true when starting the server to even get past validation to exercise this admission check for privileged pods

@liggitt Thanks for the pointer! Updated it.

**Comment by liggitt** [comment]
/lgtm
/approve

might need a rebase if https://github.com/kubernetes/kubernetes/pull/103365 beats it in

**Comment by aramase** [comment]
@liggitt Push a change to set `securityContext.AllowPrivilegeEscalation=true` in the fixtures to fix `
```
cannot set `allowPrivilegeEscalation` to false and `privileged` to true
``` 
integration test error.

I did see this as well in the last test run: `Forbidden: disallowed by cluster policy`. Looking into that now.

**Comment by liggitt** [comment]
> I did see this as well in the last test run: `Forbidden: disallowed by cluster policy`. Looking into that now.

that should be fixed by https://github.com/kubernetes/kubernetes/pull/103364#issuecomment-871617755

File: staging/src/k8s.io/pod-security-admission/test/fixtures_privileged.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+			}
+		},
+		generateFail: func(p *corev1.Pod) []*corev1.Pod {
+			p = ensureSecurityContext(p)
+			return []*corev1.Pod{
+				// privileged set to true in container
+				tweak(p, func(p *corev1.Pod) {
+					p.Spec.Containers[0].SecurityContext.Privileged = pointer.BoolPtr(true)
+					// cannot set `allowPrivilegeEscalation` to false and `privileged` to true
+					p.Spec.Containers[0].SecurityContext.AllowPrivilegeEscalation = pointer.BoolPtr(true)
```
Comment (liggitt) [inline staging/src/k8s.io/pod-security-admission/test/fixtures_privileged.go:53]: prefer setting this to nil to keep the fixtures as minimal as possible

**Comment by aramase** [comment]
> that should be fixed by #103364 (comment)

The last test run failed with that flag set in the apiserver.

File: staging/src/k8s.io/pod-security-admission/test/fixtures_privileged.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+				}),
+				// privileged set to true in init container
+				tweak(p, func(p *corev1.Pod) {
+					p.Spec.InitContainers[0].SecurityContext.Privileged = pointer.BoolPtr(true)
```
Comment (liggitt) [inline staging/src/k8s.io/pod-security-admission/test/fixtures_privileged.go:57]: set AllowPrivilegeEscalation to nil here as well

**Comment by liggitt** [comment]
> The last test run failed with that flag set in the apiserver.

are you sure? checking this branch out locally, the integration test passes once I made sure AllowPrivilegeEscalation was nil in these testcases

**Comment by aramase** [comment]
> are you sure? checking this branch out locally, the integration test passes once I made sure AllowPrivilegeEscalation was nil in these testcases

This run was with the flags as it ran after the `lgtm`:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103364/pull-kubernetes-integration/1410297911975088128.

I set `AllowPrivilegeEscalation` to nil as you suggested and ran the tests locally. The tests passed.

**Comment by liggitt** [comment]
> > are you sure? checking this branch out locally, the integration test passes once I made sure AllowPrivilegeEscalation was nil in these testcases
> 
> This run was with the flags as it ran after the `lgtm`:
> https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103364/pull-kubernetes-integration/1410297911975088128.
> 
> I set `AllowPrivilegeEscalation` to nil as you suggested and ran the tests locally. The tests passed.

figured it out... `--allow-privileged=true` plumbs to setting a global value, but it only sets the first time

other servers in the package initialized this already, so it was ineffective when running all tests in the package

to ensure the global is set the way we want, in addition to `--allow-privileged=true`, do this in the test setup:

```diff
+       "k8s.io/kubernetes/pkg/capabilities"
...
+       // ensure the global is set to allow privileged containers
+       capabilities.SetForTests(capabilities.Capabilities{AllowPrivileged: true})
```

**Comment by aramase** [comment]
> figured it out... --allow-privileged=true plumbs to setting a global value, but it only sets the first time

Ah, gotcha! Added `capabilities.SetForTests` in the tests.

**Pull request merged**